core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -1201,6 +1201,50 @@ private[chisel3] object Builder extends LazyLogging {
   }
 
   initializeSingletons()
+
+  /** The representation of the state of the [[Builder]] at a current point in
+    * time.  This is intended to capture _enough_ information to insert hardware
+    * at another point in the circuit.
+    *
+    * @see [[chisel3.Placeholder]]
+    */
+  case class State(whenStack: List[WhenContext], blockStack: List[Block], layerStack: List[layer.Layer], currentModule: Option[BaseModule])
+
+  object State {
+
+    /** Capture the current [[Builder]] state.
+      */
+    def save: State = {
+      State(Builder.whenStack, Builder.blockStack, Builder.layerStack, Builder.currentModule)
+    }
+
+    /** Set the [[Builder]] state to that provided by the parameter.
+      *
+      * @param state the state to set the [[Builder]] to
+      */
+    def restore(state: State) = {
+      Builder.whenStack = state.whenStack
+      Builder.blockStack = state.blockStack
+      Builder.layerStack = state.layerStack
+      Builder.currentModule = state.currentModule
+    }
+
+    /** Run the `thunk` with the context provided by `state`.  Save the [[Builder]]
+      * state before the thunk and restore it afterwards.
+      *
+      * @param state change the [[Builder]] to this state when running the thunk
+      * @param thunk some hardware to generate
+      * @return whatever the `thunk` returns
+      */
+    def guard[A](state: State)(thunk: => A): A = {
+      val old = save
+      restore(state)
+      val result = thunk
+      restore(old)
+      result
+    }
+
```

Comment:
Can/Should this be used in other places? Like `Module.apply`?

Reply:
Yes, probably. This is also highlighting that there is _more_ that should be saved here.

Reply:
Much more is now saved with: https://github.com/chipsalliance/chisel/pull/4628/commits/50805afb21b536af6867e532705c44867841eeae

A follow-on PR showing this being used for `Module.apply` is here: https://github.com/chipsalliance/chisel/pull/4641

I expect that this can also be used to replace the related `Definition.apply`. I haven't looked into this, yet.